merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java

```diff
@@ -734,35 +733,94 @@ public static SimulationResults computeResults(
+    final var resourceProfiles = resourceManager.computeProfiles(elapsedTime);
+    final var realProfiles = resourceProfiles.realProfiles();
+    final var discreteProfiles = resourceProfiles.discreteProfiles();
+
+    final var activityResults = computeActivitySimulationResults(engine, startTime, elapsedTime, spanInfo);
+
+    final List<Triple<Integer, String, ValueSchema>> topics = new ArrayList<>();
+    final var serializableTopicToId = new HashMap<SerializableTopic<?>, Integer>();
+    for (final var serializableTopic : serializableTopics) {
+      serializableTopicToId.put(serializableTopic, topics.size());
+      topics.add(Triple.of(topics.size(), serializableTopic.name(), serializableTopic.outputType().getSchema()));
+    }
+
+    final var spanToActivities = spanToSimulatedActivities(engine,spanInfo);
+    final var serializedTimeline = new TreeMap<Duration, List<EventGraph<EventRecord>>>();
+    var time = Duration.ZERO;
+    for (var point : this.timeline.points()) {
+      if (point instanceof TemporalEventSource.TimePoint.Delta delta) {
+        time = time.plus(delta.delta());
+      } else if (point instanceof TemporalEventSource.TimePoint.Commit commit) {
+        final var serializedEventGraph = commit.events().substitute(
+            event -> {
+              // TODO can we do this more efficiently?
+              EventGraph<EventRecord> output = EventGraph.empty();
+              for (final var serializableTopic : serializableTopics) {
+                Optional<SerializedValue> serializedEvent = trySerializeEvent(event, serializableTopic);
+                if (serializedEvent.isPresent()) {
+                  // If the event's `provenance` has no simulated activity id, search its ancestors to find the nearest
+                  // simulated activity id, if one exists
+                  if (!spanToActivities.containsKey(event.provenance())) {
+                    var spanId = Optional.of(event.provenance());
+
+                    while (true) {
+                      if (spanToActivities.containsKey(spanId.get())) {
+                        spanToActivities.put(event.provenance(), spanToActivities.get(spanId.get()));
+                        break;
+                      }
+                      spanId = engine.getSpan(spanId.get()).parent();
+                      if (spanId.isEmpty()) {
+                        break;
+                      }
+                    }
+                  }
+                  var activitySpanID = Optional.ofNullable(spanToActivities.get(event.provenance()).id());
+                  output = EventGraph.concurrently(
+                      output,
+                      EventGraph.atom(
+                          new EventRecord(serializableTopicToId.get(serializableTopic),
+                                          activitySpanID,
+                                          serializedEvent.get())));
+                }
+              }
```

Comment:
I've extracted this code into it's own method rather than repeating it in both `computeResults`. I've left it as it's own commit near the end to avoid rebase headaches.